File: src/main/java/com/pancisin/bookster/models/MetaField.java
package com.pancisin.bookster.models;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.pancisin.bookster.models.enums.MetaType;

@Entity
@Table(name = "meta_fields")
public class MetaField {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Column
	private String name;

	@Enumerated(EnumType.STRING)
	private MetaType type;

	@Column
	private String description;

	@JsonIgnore
	@OneToMany(mappedBy = "field")
	private List<MetaValue> values;

	@ElementCollection
	@Column
	private List<String> options;

	@Column
	private boolean optional;

	@Column
	private int ordering;
	
	@Override
	public String toString() {
		return this.id.toString();
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public List<String> getOptions() {
		return options;
	}

	public void setOptions(List<String> options) {
		this.options = options;
	}

	public boolean isOptional() {
		return optional;
	}

	public void setOptional(boolean optional) {
		this.optional = optional;
	}

	public Long getId() {
		return id;
	}

	public MetaType getType() {
		return type;
	}

	public void setType(MetaType type) {
		this.type = type;
	}

	public List<MetaValue> getValues() {
		return values;
	}

	public int getOrdering() {
		return ordering;
	}

	public void setOrdering(int ordering) {
		this.ordering = ordering;
	}
}

File: src/main/kotlin/com/pancisin/bookster/model/MetaField.kt
package com.pancisin.bookster.model

import javax.persistence.Column
import javax.persistence.ElementCollection
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.OneToMany
import javax.persistence.Table

import com.fasterxml.jackson.annotation.JsonIgnore
import com.pancisin.bookster.model.enums.MetaType

@Entity @Table(name = "meta_fields")
data class MetaField(
  @Id @GeneratedValue(strategy = GenerationType.AUTO)
  val id: Long? = null,

  @Column
  var name: String? = null,

  @Enumerated(EnumType.STRING)
  var type: MetaType? = null,

  @Column
  var description: String? = null,

  @JsonIgnore @OneToMany(mappedBy = "field")
  var values: List<MetaValue>? = null,

  @ElementCollection @Column
  var options: List<String>? = null,

  @Column
  var optional: Boolean = false,

  @Column
  var ordering: Int = 0
) {
  override fun toString(): String {
    return this.id.toString()
  }
}
